### PR TITLE
Make DebugLogger and EventLogLogger externally configurable

### PR DESCRIPTION
**Pull request by cesarbs**
Addresses #427 and more.

Right now, only `ConsoleLogger` can be configured from an `IConfiguration` instance. This change makes `DebugLogger` and `EventLogLogger` configurable via `IConfiguration`, also supporting automatic reloading.

The changes are mostly additive, but there's a breaking change in that I've removed `ConfigurationConsoleLoggerSettings`, `ConsoleLoggerSettings` and `IConsoleLoggerSettings`.

To prevent duplicating logic across all three loggers, I've replaced `ConfigurationConsoleLoggerSettings` with `ConfigurableLoggerSettings` and put that in `Microsoft.Extensions.Logging.Abstractions`.

Along with that, I've introduced an abstract type named `ConfigurableLoggerProvider`, which `ConsoleLoggerProvider`, `DebugLoggerProvider` and `EventLogLoggerProvider` now subclass. `ConfigurableLoggerProvider` contains the configuration logic which used to reside in `ConsoleLoggerProvider`.

I don't think the filter logger and the TraceSource logger fit into this new model, so I've left them as is.

A few points pending:

* For now I've put `ConfigurableLoggerProvider` and `ConfigurableLoggerSettings` in `Microsoft.Exntesions.Logging.Abstractions`, but I think they might warrant their own package (`Microsoft.Extensions.Logging.Configuration`?). I also thought of putting them in `Microsoft.Extensions.Logging`, but that would change other loggers' dependencies from depending on that package instead of only depending on the `.Abstractions` one.

* `DebugLogger` and `EventLogLogger` now have an `IncludeScopes` property, but it doesn't do anything right now.

* More tests.

cc @kichalla @BrennanConroy @lodejard @davidfowl @Eilon @muratg

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.Abstractions/ConfigurableLoggerProvider.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                throw new ArgumentNullException(nameof(filter));
+            }
+
+            _filter = filter;
+            _settings = new ConfigurableLoggerSettings(new ConfigurationBuilder()
+                .Add(new MemoryConfigurationSource
+                {
+                    InitialData = new Dictionary<string, string>
+                    {
+                        ["IncludeScopes"] = "false"
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.Logging.Abstractions/ConfigurableLoggerProvider.cs]: Shouldn't you be using the value of `includeScopes` here?

File: src/Microsoft.Extensions.Logging.Debug/DebugLoggerFactoryExtensions.cs [hunk 2]
```diff
@@ -24,10 +26,24 @@ public static ILoggerFactory AddDebug(this ILoggerFactory factory)
         /// Adds a debug logger that is enabled as defined by the filter function.
         /// </summary>
         /// <param name="factory">The extension method argument.</param>
+        /// <param name="includeScopes">A value which indicates whether log scope information should be displayed
+        /// in the output.</param>
+        public static ILoggerFactory AddDebug(this ILoggerFactory factory, bool includeScopes)
+        {
+            factory.AddDebug((n, l) => l >= LogLevel.Information, false);
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.Logging.Debug/DebugLoggerFactoryExtensions.cs]: Use `includeScopes`?

File: src/Microsoft.Extensions.Logging.Debug/DebugLoggerFactoryExtensions.cs [hunk 3]
```diff
@@ -24,10 +26,24 @@ public static ILoggerFactory AddDebug(this ILoggerFactory factory)
...
+        /// </summary>
+        /// <param name="factory">The extension method argument.</param>
         /// <param name="filter">The function used to filter events based on the log level.</param>
-        public static ILoggerFactory AddDebug(this ILoggerFactory factory, Func<string, LogLevel, bool> filter)
+        /// <param name="includeScopes">A value which indicates whether log scope information should be displayed
+        /// in the output.</param>
+        public static ILoggerFactory AddDebug(this ILoggerFactory factory, Func<string, LogLevel, bool> filter, bool includeScopes)
         {
-            factory.AddProvider(new DebugLoggerProvider(filter));
+            factory.AddProvider(new DebugLoggerProvider(filter, false));
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.Logging.Debug/DebugLoggerFactoryExtensions.cs]: Again, use `includeScopes`?

**Comment by Eilon** [comment]
Moving this to 1.0.1 as well.

**Comment by muratg** [comment]
@cesarbs what are we doing with this one?

**Comment by cesarbs** [comment]
@muratg I'll get back to it. I think I'll go over another "full iteration" on the idea since I'm not happy with a number of things in this first attempt.

**Comment by muratg** [comment]
@cesarbs this one good to close then?

**Pull request closed without merging**